Upgrades to contact messages

joemull opened the pull request:
Closes #4138.
Closes #1720.

~Please also review and merge docs PR: https://github.com/openlibhums/memory-alpha/pull/155.~

## Notes

* Closes the linked issues
* Connects the contact person and contact message models to account, so that we do not have names and emails outside of the Account object
* Renames the contact person and contact message models to make them easier to read and work with
* Removes stored client IPs to comply with GDPR
* Adds the Django auto-documentation views, which are helpful when searching for template tags and filters, and which show you view and url pairs
* Documents a few things
* Tests for the contact manager and the press and journal contact pages

## Screenshots for contact system

<img width="1149" height="749" alt="Screenshot from 2025-07-21 17-32-46" src="https://github.com/user-attachments/assets/cf1342ba-ab41-4af6-bf19-94b86f88ec9f" />
<img width="1125" height="520" alt="Screenshot from 2025-07-21 17-33-10" src="https://github.com/user-attachments/assets/c9e51dc6-c63b-44a1-8484-ce867d2b73b5" />
<img width="1125" height="685" alt="Screenshot from 2025-07-21 17-33-20" src="https://github.com/user-attachments/assets/38e03aa3-ab13-47b7-b9ba-f87d190b05fc" />
<img width="1125" height="498" alt="Screenshot from 2025-07-21 17-33-46" src="https://github.com/user-attachments/assets/059de799-8e05-445a-a2ac-3ae1707e6e2f" />
<img width="1125" height="763" alt="Screenshot from 2025-07-21 17-34-03" src="https://github.com/user-attachments/assets/e1d7fe6d-fa96-4015-b744-c43e1b8b2a8f" />
<img width="1125" height="644" alt="Screenshot from 2025-07-21 17-34-14" src="https://github.com/user-attachments/assets/0e5e104f-38f6-4c52-bfb6-5be674831767" />

## Screenshots for admin docs feature

I enabled this in the process of doing this work. It is enormously helpful to view a comprehensive list of things in Janeway sometimes.

<img width="1306" height="1164" alt="Screenshot from 2025-07-21 17-35-02" src="https://github.com/user-attachments/assets/d02dd4bf-688f-4625-bba7-80aa40c63aaa" />
<img width="1306" height="1164" alt="Screenshot from 2025-07-21 17-35-18" src="https://github.com/user-attachments/assets/e36c3d39-600c-4efb-ada1-bf23e935eb9c" />
<img width="1306" height="1164" alt="Screenshot from 2025-07-21 17-40-34" src="https://github.com/user-attachments/assets/380fbef2-0e46-4f6d-836e-77b1deab8e7b" />

joemull committed:
docs #4138 Add to view test docs

joemull committed:
docs #4138 Turn on Django admin docs

joemull committed:
docs #4138 Document sequence fields and email logging

joemull committed:
feat #4138 Upgrade and refactor contact system

joemull committed:
fix #4138: Fix failing tests

ajrbyers reviewed (changes requested):
I've added a few comments inline for you to go over, however, I have a fundamental question about this implementation. We have a contact form to obscure email addresses from being scraped. If we require a user to create a URL with the users email address in it we may be defeating this purpose. I think the account UUID field is the least painful option here. The Contacts page could present contact link that a user could simply copy and paste so they don't have to form it themselves?

Open to thoughts on this.

joemull committed:
feat #4138 Use account uuid not email to match recipient

joemull committed:
feat #4138 Add button to copy link, and other UI improvements

joemull committed:
feat #4138 Avoid relying on model manager in migration

joemull commented:
> I've added a few comments inline for you to go over, however, I have a fundamental question about this implementation. We have a contact form to obscure email addresses from being scraped. If we require a user to create a URL with the users email address in it we may be defeating this purpose. I think the account UUID field is the least painful option here. The Contacts page could present contact link that a user could simply copy and paste so they don't have to form it themselves?
> 
> Open to thoughts on this.

All good suggestions, thanks! Updated with some changes:

* Used uuid as suggested
* Added link copy button as suggested
* Went ahead and standardised other buttons and confirm delete behavior for consistency
* Fixed other display issues
* Updated migration to fix bug, and tested with OLH data

joemull commented:
Note: Going to rebase this as there are conflicts keeping the CI job from running.

joemull committed:
feat #4138 Change contact URLs to new press view

mauromsl commented on src/core/logic.py:
```diff
@@ -1221,3 +1221,58 @@ def create_organization_name(request):
+            account = models.Account.objects.get(
+                contactperson__in=contact_people,
+                uuid=recipient_uuid,
+            )
```
If I know the UUID of another account, I think this potentially allows an attacker to get a contact from from a different site. Instead, and since `request` is an argument to this function, we can flip the query to retrieve a `Contact` person and match against the `request.site_type`

mauromsl reviewed (changes requested):
a few comments inline

joemull commented on src/core/views.py:
```diff
@@ -1870,117 +1878,184 @@ def article_image_edit(request, article_pk):
+        content_type=request.model_content_type,
+        object_id=request.site_type.pk,
+    )
+    contact_people = request.site_type.contact_people
```
It is used in the context by a sub-template, as above. I've added `with contact_people=contact_people` in the template inclusions to show the trail more clearly.

joemull reviewed (commented):
OK I've written some replies and changed a few things--commits incoming once I test it on OLH data again.

joemull committed:
feat #4138 Make data migration more robust

joemull committed:
feat #4138 Make variables in included templates more explicit

mauromsl reviewed (changes requested):
Thanks for the clarifications Joe, I've added one comment with some final suggestions inline

ajrbyers commented on src/core/models.py:
```diff
@@ -1970,32 +1975,69 @@ class Contacts(models.Model):
+
+class ContactMessage(models.Model):
+    account = models.ForeignKey(
+        Account,
```
I prefer option b. Contact messages are emails like any other system communication, so they should be stored in logs for consistency.

The counter-intuitiveness issue is actually an opportunity - we can use this to begin introducing non-article linked logs. Start simple: add a way to view logs that aren't attached to articles, filterable by type. Contact messages become the first use case, and we build the minimal interface needed to access them, and a link to that page can apply a filter to show contact messages for that particular journal.

joemull committed:
feat #4138 Store contact messages as log entries and create UI for them

joemull committed:
feat #4138 Document upgraded contact system

joemull committed:
feat #4138 Fix rebase bug and update docs

mauromsl reviewed (changes requested):
a few comments inline around the complexity of email comparisons and related data migrations.

joemull committed:
feat #4138 Changes requested in review

joemull committed:
feat #4138 Add manager migration and fix GFK error in data migration

mauromsl commented on src/core/migrations/0111_rename_contacts_contactperson_and_more.py:
```diff
@@ -0,0 +1,267 @@
+                username=normalized_email,
+                first_name=" ".join(contact_person.name.split()[:-1]),
+                last_name=contact_person.name.split()[-1],
+            )
```
I'm thinking cases I've seen where

- email has been inserted without normalization (due to a db operation or an import routine ignoring application-level checks)
- email and account are different to each other (e.g. account created with one email address, like a dummy email, and then later have their email address added).

for data migration code specifically, we always want to be overzealous and protect against edge-cases that are possible at the database level. 

In fact, I just thought of another case where we'll need to protect against `MultipleObjectsReturned`), so it is worth handling it as well.

mauromsl commented on src/core/forms/forms.py:
```diff
@@ -106,6 +104,37 @@ class Meta:
+        label=_("Who would you like to contact?"),
+    )
+    sender = forms.EmailField(
+        max_length=200,
```
Sorry, I meant the `max_length`, by using a constant, so that if one changes, the other does too. (since we copy from one to the other)

mauromsl reviewed (changes requested):
replied to comments, we are nearly there

joemull committed:
feat #4138 Avoid unforeseen migration errors; optimize; fix date bug

joemull commented:
OK @mauromsl, I've addressed your comments. I've also tested it with a larger database and optimized the db access in the migration so it's quicker. There was also a bug with log entry dates because `auto_now_add` can't be manually overridden, so I made another change to that model.

mauromsl reviewed (changes requested):
Couple changes suggested inline for the data migration

joemull commented on src/core/migrations/0111_rename_contacts_contactperson_and_more.py:
```diff
@@ -1,90 +1,116 @@
+    Journal = apps.get_model("journal", "Journal")
+    Role = apps.get_model("core", "Role")
+    AccountRole = apps.get_model("core", "AccountRole")
+    for contact_person in ContactPerson.objects.all():
```
I don't anticipate any contact people would have accounts when this is run, but it can't hurt.

joemull commented on src/core/migrations/0111_rename_contacts_contactperson_and_more.py:
```diff
@@ -112,21 +138,23 @@ def move_contact_message_to_log_entry(apps, schema_editor):
+        # Link up the addressees with their log entries
+        addressees_to_create = []
+        for i, addressee in enumerate(addressees_to_link_up):
+            addressee.log_entry = log_entries[i]
```
OK, I will give this a try. By the way, I believe this might be the bug behind #5248, because I had to use a similar strategy to link the different pieces of the ROR data model together without it taking an hour to run.

joemull replied:
Done and tested on MySQL.

joemull committed:
feat #4138 Support MySQL and improve idempotency

joemull committed:
docs #4138 Fix typos in docs

joemull commented:
Need to resolve #5259 before I can finish testing this. Setting status as blocked.

joemull committed:
feat #4138 Fix migration graph

joemull committed:
feat #4138 Respect From field even for logged-in users

joemull committed:
feat #4138 Tiny edit

joemull commented:
> Need to resolve #5259 before I can finish testing this. Setting status as blocked.

- Rebased in order to get the MySQL bug fix.
- Tested manually on MySQL. 
- Made a small adjustment for logged-in users filling out contact form to match user expectation.

mauromsl commented on src/core/logic.py:
```diff
@@ -1370,3 +1371,78 @@ def resolve_alt_text_target(request):
+        "actor_email": contact_form.cleaned_data["sender"],
+    }
+
+    # Avoid sending request.user through the logging system,
+    # because the contact page is public-facing and logged-in users
+    # will not expect it to record their account email as the LogEntry.actor
+    # rather than the email they put in the From field.
+    request.user = None
```
This is a dangerous proposition, as `request.user` is global for the entire request lifecycle. I'd recommend either handling this downstream or changing the code to take an explicit user that is set to `None` for this instance

mauromsl commented:
@joemull I had to revert the merge of this PR, as I was a little to eager to merge and only realised there is a problem after merging.

Since we can't reopen the PR, would you mind opening a new one and linking them through

https://github.com/openlibhums/janeway/pull/4845#discussion_r3461181260

Apologies for the confusion

mauromsl commented:
now closed by #5367
